### PR TITLE
ensure consistent structure for groupby on index and column

### LINKED ISSUE #17681: groupby on index and columns mixed: strange behaviour
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
df1 = pd.DataFrame([[1, 2, 3]], columns=['a', 'b', 'c']).set_index('a')
df2 = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=['a', 'b', 'c']).set_index('a')
df3 = pd.DataFrame([[1, 2, 3], [4, 5, 6], [7, 8, 9]], columns=['a', 'b', 'c']).set_index('a')

print(df1.groupby(['a', 'c']).sum())
print('==============')
print(df2.groupby(['a', 'c']).sum())
print('==============')
print(df3.groupby(['a', 'c']).sum())
# output:
#    b
# a c   
# 1 3  2
# ==============
#    b  c
# a  2  3
# c  5  6
# ==============
#      b
# a c   
# 1 3  2
# 4 6  5
# 7 9  8
```
#### Problem description

This is a problem because the output structure depends on the number of rows in a dataframe. I figured out that this issue can be solved by calling reset_index() on the dataframe before the groupby, but this issue has caused unexpected troubles for our production framework. 
Here the only **exception** is when the input dataframe has **2 rows**.

#### Expected Output

```
     b
a c   
1 3  2
==============
     b
a c   
1 3  2
4 6  5
==============
     b
a c   
1 3  2
4 6  5
7 9  8
```

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Darwin
OS-release: 16.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.3
pytest: 3.1.3
pip: 9.0.1
setuptools: 33.1.1.post20170320
Cython: 0.26
numpy: 1.13.1
scipy: 0.19.1
xarray: None
IPython: 6.1.0
sphinx: None
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
sqlalchemy: 1.1.13
pymysql: 0.7.9.None
psycopg2: 2.7.3 (dt dec pq3 ext lo64)
jinja2: 2.9.5
s3fs: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by mjarosie**
- [x] closes #17681
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/tests/groupby/test_grouping.py [hunk 1]
```diff
@@ -641,6 +642,27 @@ def test_groupby_level_index_value_all_na(self):
...
+            pd.DataFrame([[1, 2, 3]], columns=["a", "b", "c"]).set_index("a"),
+            pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=["a", "b", "c"]).set_index(
+                "a"
+            ),
+            pd.DataFrame(
+                [[1, 2, 3], [4, 5, 6], [7, 8, 9]], columns=["a", "b", "c"]
+            ).set_index("a"),
+        ],
+    )
+    @pytest.mark.parametrize(
```
Comment (jreback) [inline pandas/tests/groupby/test_grouping.py]: can you use the fixture all_numeric_reductions (you can simply replace method with this)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_grouping.py (hunk 1 above)
Comment (mjarosie) [reply pandas/tests/groupby/test_grouping.py]: I've just checked, if I remove the second pytest parametrisation and use `all_numeric_reductions` (from `pandas.conftest`) as the second parameter the test will throw an error of `AttributeError: 'DataFrameGroupBy' object has no attribute 'kurt'`.

Comment (WillAyd) [reply pandas/tests/groupby/test_grouping.py]: I think can also use `reduction_func` from the groupby contest. Might need to consolidate those fixtures at a later point

Comment (mjarosie) [reply pandas/tests/groupby/test_grouping.py]: In that case tests fail for `ngroup` and `nth` reductions: `AssertionError: assert FrozenList(['a']) == FrozenList(['a', 'c'])` and `TypeError: nth() missing 1 required positional argument: 'n'` respectively.

Comment (WillAyd) [reply pandas/tests/groupby/test_grouping.py]: So the test in #29124 is already doing this with all of the groupby funcs though with a different intent (duplicate column labels). Maybe after that gets merged we come back to this and see if this makes sense to fit into that test as well via some sort of parametrization; not sure it's worth doing as a one-off

**Comment by WillAyd** [comment]
Thanks for the PR! But I think this was already covered in #29124

**Pull request closed without merging**